src/upload-lighthouse/index.ts

```diff
@@ -0,0 +1,185 @@
+  htmlPath: string;
+  jsonPath: string;
+  summary: Summary;
+}
+
+interface JsonRun {
+  /* eslint-disable-next-line  @typescript-eslint/no-explicit-any */
+  [k: string]: any;
+}
+
+const DB_NAME = `lighthouse`;
+const PR_COLL_NAME = `pr_reports`;
+const MAIN_COLL_NAME = `main_reports`;
+
+const getEmptySummary = (): Summary => ({
+  seo: 0,
+  performance: 0,
+  'best-practices': 0,
+  pwa: 0,
+  accessibility: 0,
+});
+
+const getAverageSummary = (manifests: Manifest[]): Summary => {
+  const summary = getEmptySummary();
+  for (const property of summaryProperties) {
+    summary[property] =
+      manifests.reduce((acc, cur) => acc + cur.summary[property], 0) /
+      manifests.length;
+  }
+  return summary;
+};
+
+const getRuns = async (
+  manifests: Manifest[],
+): Promise<{ jsonRuns: JsonRun[]; htmlRuns: string[] }> => {
+  const jsonRuns = [];
+  const htmlRuns = [];
+
+  for (const manifest of manifests) {
+    jsonRuns.push(
+      JSON.parse((await readFileAsync(manifest.jsonPath)).toString()),
+    );
+
+    htmlRuns.push((await readFileAsync(manifest.htmlPath)).toString());
+  }
+
+  await Promise.all(jsonRuns);
+  await Promise.all(htmlRuns);
+  return { jsonRuns, htmlRuns };
+};
+
+const sortAndReadRuns = async (
+  manifests: Manifest[],
+): Promise<
+  { jsonRuns: JsonRun[]; htmlRuns: string[]; summary: Summary; url: string }[]
+> => {
+  const runs: {
+    jsonRuns: JsonRun[];
+    htmlRuns: string[];
+    summary: Summary;
+    url: string;
+  }[] = [];
+  const uniqueUrls = new Set(manifests.map(manifest => manifest.url));
+
+  for (const url of uniqueUrls) {
+    const manifestsForUrl = manifests.filter(manifest => manifest.url === url);
+    const summary = getAverageSummary(manifestsForUrl);
+    const { jsonRuns, htmlRuns } = await getRuns(manifestsForUrl);
+    runs.push({ jsonRuns, htmlRuns, summary, url });
+  }
+
+  return runs;
+};
+
+async function main(): Promise<void> {
+  const commitHash = github.context.sha;
+  const author = github.context.actor;
+  const commitMessage = process.env.COMMIT_MESSAGE || '';
+  const commitTimestamp = process.env.COMMIT_TIMESTAMP || '';
+  const project = process.env.PROJECT_TO_BUILD || '';
+  const branch = process.env.BRANCH_NAME || '';
+
+  try {
+    const outputsFile = (
+      await readFileAsync('./lhci/manifest.json')
+    ).toString();
+
+    const manifestsOfLighthouseRuns: Manifest[] = JSON.parse(outputsFile);
+
+    const [desktopRunManifests, mobileRunManifests] =
+      manifestsOfLighthouseRuns.reduce(
+        (acc, cur) => {
+          if (cur.url.includes('?desktop')) acc[0].push(cur);
+          else acc[1].push(cur);
+          return acc;
+        },
+        [[], []] as Manifest[][],
+      );
+
+    const desktopRuns = await sortAndReadRuns(desktopRunManifests);
+    const desktopRunDocuments = [];
+
+    for (const desktopRun of desktopRuns) {
+      const { htmlRuns, jsonRuns, summary, url } = desktopRun;
+      desktopRunDocuments.push({
```

Comment:
can we pre-create the shared properties and add the `type` (vs currently we are creating same one twice)

Reply:
ie. can run through all of `manifestsOfLighthouseRuns` and determine type when reading the manifest.url